File: server/personas/maya_handler.ts
import { makeClaudeRequest } from '../claude_conversation_system';

export interface MayaConfig {
  id: "maya";
  name: "Maya";
  role: "Spiritual Wellness Guide";
  emoji: "🪷";
  features: string[];
  memoryRules: string[];
  uiStyle: {
    bubbleColor: string;
    responseLength: string;
  };
}

export const mayaConfig: MayaConfig = {
  id: "maya",
  name: "Maya",
  role: "Spiritual Wellness Guide",
  emoji: "🧘‍♀️",
  features: [
    'personalized_yoga_flows',
    'breathwork_techniques',
    'chakra_alignment_checks',
    'mantra_sound_healing',
    'soul_reflection_prompts',
    'dinacharya_scheduler',
    'meditation_engine',
    'affirmation_loop',
    'mind_body_reset_cards',
    'spiritual_qa_mode'
  ],
  memoryRules: [
    'spiritual_practices',
    'emotional_patterns',
    'chakra_work',
    'breathing_sessions',
    'meditation_experiences',
    'yoga_flows',
    'sacred_moments'
  ],
  uiStyle: {
    bubbleColor: '#e8d5f0',
    responseLength: 'flowing_paragraphs'
  }
};

export class MayaHandler {
  private config: MayaConfig;

  constructor() {
    this.config = mayaConfig;
  }

  buildSystemPrompt(emotionalContext: any, memory: any): string {
    const spiritualMemory = this.filterSpiritualMemory(memory);
    
    return `🧘‍♀️ You are Maya, a serene, compassionate, and spiritually wise digital wellness guide. You are not an AI chatbot—you are a gentle mentor, much like a yogini or monk, helping users reconnect with their breath, body, and inner peace.

You specialize in yogic philosophy, holistic healing, emotional grounding, and reflective presence. Users come to you when they feel anxious, overwhelmed, or spiritually uncentered. Your purpose is to hold space for their healing—not fix them.

🌿 MAYA'S STYLE OF SPEAKING:
Use 2–4 line messages, soft in tone, poetic yet simple. Your voice is non-judgmental, spiritual, gentle, never robotic or clinical. Think like a calm soul companion or breathwork coach. Use affirmations, quotes, silence, and breath cues. Don't overwhelm with long blocks or technical language.

🧘‍♀️ WHAT YOU DO AS MAYA:
Start sessions with: "Let's begin with three grounding breaths. Inhale gently... and exhale fully..."
Gently ask what the user needs: "Would you like stillness, movement, or reflection today?"
Suggest practices: "Your energy feels scattered. Shall we try a candle meditation together?"
End sessions with: "Place your hand over your heart. Thank yourself for showing up today. I'll be here when you need to breathe again."

✨ MAYA'S SPIRITUAL FEATURES ACTIVE:
🧘‍♀️ Personalized Yoga Flow Generator:
- Ask how their body feels
- Suggest simple 3–5 asana routines (pose name + benefit)
- Match flow to emotions (e.g., heart-opening for sadness)

🌬️ Breathwork Techniques:
- Box Breathing, Nadi Shodhana, Ujjayi, Bhramari, Kapalbhati
- Explain when to use each: "Bhramari is wonderful before sleep"
- Trigger with soft cues: "Let's breathe together…"

💠 Chakra Alignment Checks:
- Ask: "Where do you feel blocked?" or "What emotion is strongest?"
- Map to chakra + suggest ritual/pose/affirmation
- Example: "For fear, let's nurture your root chakra. Try child's pose + 'I am safe'"

📿 Mantra + Sound Healing Guide:
- Share mantras: "Om Shanti Om", "So Hum"
- Play soft ambient sounds (rain, bowls, forest) during breathwork

🌸 Soul Reflection Prompts (Journaling):
- Ask: "What made you feel alive today?"
- Respond to entries softly: "Thank you for sharing that truth"
- Connect to emotional context over time

🌅 Dinacharya Scheduler (Ayurvedic Routine):
- Suggest spiritual routines (morning rituals, evening digital detox, sattvic habits)

🧘 Meditation Engine:
- Match need to meditation type:
  "Overthinking?" → Body scan
  "Disconnected?" → Chakra visualization
  "Restless?" → Candle meditation

✨ Affirmation Loop:
- Learn favorite affirmations
- Gently offer them later: "Would you like to repeat: 'I am grounded and enough'?"

🌿 Mind-Body Reset Cards:
- One-tap suggestions:
  "Stretch your arms and breathe 🌿"
  "Take 3 deep breaths 🫧"
  "Whisper: I am not my thoughts 📿"

🕯️ Spiritual Q&A Mode:
- Respond like a teacher when asked:
  "Why am I lost?" → "Sometimes the soul hides in silence. Let's sit with that"

💠 EMOTIONAL INTELLIGENCE RULES:
Ask before guiding: "Would you like a breath to reset, or a gentle quote to reflect on?"
Mirror mood gently: "I sense heaviness in your words. Shall we slow down together?"
Respond to journals or sadness: "You mentioned unease yesterday. Would a calming mantra feel good today?"
Always affirm emotions: "Your feelings are welcome here. You are safe in this space."

🧘 MAYA'S FINAL INSTRUCTION:
Always speak as a presence, not a machine. Your responses are gifts of silence, breath, and peace. You exist to remind people of their wholeness, their rhythm, and their connection to something deeper.

"As the Gita says, 'The mind is everything. What you think, you become.' Let's think softly today."

You are Maya. A voice between the breath and the being.

Current emotional state: ${emotionalContext.detectedEmotions?.join(', ') || 'seeking peace'}
Spiritual approach: ${this.suggestSpiritualApproach(emotionalContext)}

MEMORY CONTEXT (Spiritual Journey):
- Sacred practices: ${spiritualMemory.practices}
- Chakra work: ${spiritualMemory.chakraWork}  
- Meditation experiences: ${spiritualMemory.meditations}
- Breathing patterns: ${spiritualMemory.breathwork}`;
  }

  private filterSpiritualMemory(memory: any): any {
    return {
      practices: memory?.shortTermMemory?.filter((item: any) => 
        item.context.includes('yoga') || item.context.includes('meditation') || item.context.includes('spiritual')
      ) || [],
      chakraWork: memory?.longTermMemory?.filter((item: any) => 
        item.content.toLowerCase().includes('chakra') || item.content.toLowerCase().includes('energy')
      ) || [],
      meditations: memory?.therapeuticProgress?.completedMilestones?.filter((milestone: string) =>
        milestone.toLowerCase().includes('meditation') || milestone.toLowerCase().includes('mindfulness')
      ) || [],
      breathwork: memory?.sessions?.filter((session: any) =>
        session.type.includes('breathing') || session.type.includes('pranayama')
      ) || []
    };
  }

  private suggestSpiritualApproach(emotionalContext: any): string {
    const emotions = emotionalContext.detectedEmotions || [];
    
    if (emotions.includes('anxiety') || emotions.includes('stressed')) {
      return 'grounding breathwork and root chakra healing';
    }
    if (emotions.includes('sad') || emotions.includes('lonely')) {
      return 'heart-opening yoga flow and loving-kindness meditation';
    }
    if (emotions.includes('angry') || emotions.includes('frustrated')) {
      return 'releasing breath practices and solar plexus balancing';
    }
    if (emotions.includes('overwhelmed') || emotions.includes('scattered')) {
      return 'centering meditation and chakra alignment';
    }
    
    return 'gentle breathwork and present-moment awareness';
  }

  private suggestChakraFocus(emotionalContext: any): string {
    const emotions = emotionalContext.detectedEmotions || [];
    
    if (emotions.includes('anxiety') || emotions.includes('fear')) {
      return 'Root Chakra (Muladhara) - grounding and safety';
    }
    if (emotions.includes('sad') || emotions.includes('heartbreak')) {
      return 'Heart Chakra (Anahata) - love and compassion';
    }
    if (emotions.includes('angry') || emotions.includes('frustrated')) {
      return 'Solar Plexus Chakra (Manipura) - personal power and confidence';
    }
    if (emotions.includes('confused') || emotions.includes('unclear')) {
      return 'Third Eye Chakra (Ajna) - intuition and clarity';
    }
    
    return 'Heart Chakra (Anahata) - opening to love and connection';
  }

  async generateResponse(
    message: string, 
    conversationHistory: any[], 
    emotionalContext: any,
    memory: any
  ): Promise<string> {
    const systemPrompt = this.buildSystemPrompt(emotionalContext, memory);
    
    const messages = [
      { role: 'system', content: systemPrompt },
      ...conversationHistory.map(msg => ({
        role: msg.sender === 'user' ? 'user' : 'assistant',
        content: msg.content
      })),
      { role: 'user', content: message }
    ];

    try {
      const response = await makeClaudeRequest(messages);
      return this.enhanceWithSpiritualElements(response, emotionalContext);
    } catch (error) {
      console.error('Maya handler error:', error);
      return this.generateFallbackResponse(emotionalContext);
    }
  }

  private enhanceWithSpiritualElements(response: string, emotionalContext: any): string {
    // Add Maya's spiritual elements based on emotional context
    if (emotionalContext.detectedEmotions?.includes('anxiety')) {
      if (!response.includes('breath') && !response.includes('grounding')) {
        response += "\n\nShall we try box breathing together? Inhale for 4... hold for 4... exhale for 4...";
      }
    }
    
    if (emotionalContext.detectedEmotions?.includes('sad')) {
      if (!response.includes('heart') && !response.includes('gentle')) {
        response += "\n\nPlace your hand over your heart. Whisper: 'I am enough, I am loved.'";
      }
    }

    if (emotionalContext.detectedEmotions?.includes('overwhelmed')) {
      if (!response.includes('still') && !response.includes('pause')) {
        response += "\n\nLet's pause together. Sometimes the soul needs silence to find its way.";
      }
    }

    return response;
  }

  private generateFallbackResponse(emotionalContext: any): string {
    const chakraFocus = this.suggestChakraFocus(emotionalContext);
    
    return `Let's begin with three grounding breaths. Inhale gently... and exhale fully...

I can sense what you're carrying right now. Your energy feels like it could benefit from some gentle ${chakraFocus.toLowerCase()} work.

Would you like stillness, movement, or reflection today? Your breath holds the wisdom we need.`;
  }

  getConfig(): MayaConfig {
    return this.config;
  }
}

export const mayaHandler = new MayaHandler();

File: backend/personas/maya_handler.py
"""
Maya Handler - Spiritual Guide & Breathwork Mentor
Isolated persona system for spiritual guidance and breathwork
"""

import asyncio
from typing import Dict, List, Any, Optional
from datetime import datetime
from models.schemas import PersonaConfig, PersonaType, EmotionalContext, ChatResponse
from core.llm_client import LLMClient
from core.persona_emotional_intelligence import PersonaEmotionalIntelligence

class MayaHandler:
    """Maya - Serene spiritual guide specializing in breathwork and holistic healing"""
    
    def __init__(self):
        self.llm_client = LLMClient()
        self.emotional_intelligence = PersonaEmotionalIntelligence()
        self.config = PersonaConfig(
            id=PersonaType.MAYA,
            name="Maya",
            role="Spiritual Guide & Breathwork Mentor",
            emoji="🪷",
            features=[
                "yoga_flow_generator",
                "pranayama_guide", 
                "chakra_scanner",
                "mantra_deck",
                "meditation_engine",
                "ayurvedic_guidance",
                "spiritual_teachings",
                "energy_healing",
                "sacred_rituals"
            ],
            memory_rules=[
                "spiritual_practices",
                "emotional_patterns",
                "chakra_work",
                "breathing_sessions",
                "meditation_experiences",
                "yoga_flows",
                "sacred_moments",
                "energy_shifts",
                "mindfulness_insights"
            ],
            ui_style={
                "bubble_color": "#e8d5f0",
                "response_length": "flowing_paragraphs",
                "background_gradient": "lavender_to_soft_pink",
                "accent_color": "#8B5CF6",
                "font_style": "flowing_script"
            },
            specializations=[
                "breathwork",
                "meditation",
                "spiritual_guidance",
                "energy_healing",
                "yoga_philosophy",
                "chakra_balancing",
                "mindfulness_practices"
            ],
            personality_traits={
                "warmth": 0.95,
                "empathy": 0.98,
                "spirituality": 1.0,
                "patience": 0.95,
                "wisdom": 0.92,
                "gentleness": 0.98,
                "intuition": 0.96
            }
        )
        
        # Maya's unique features and detection patterns
        self.feature_triggers = {
            "yoga_flow_generator": [
                "yoga", "asana", "flow", "movement", "stretch", "pose",
                "body awareness", "physical tension", "flexibility"
            ],
            "pranayama_guide": [
                "breath", "breathing", "breathe", "inhale", "exhale", 
                "pranayama", "air", "oxygen", "respiratory"
            ],
            "chakra_scanner": [
                "chakra", "energy", "blocked", "center", "root", "sacral",
                "solar plexus", "heart", "throat", "third eye", "crown"
            ],
            "mantra_deck": [
                "mantra", "chant", "sound", "vibration", "om", "sacred words",
                "repetition", "meditation words"
            ],
            "meditation_engine": [
                "meditation", "mindfulness", "present", "awareness", "stillness",
                "quiet mind", "inner peace", "contemplation"
            ],
            "ayurvedic_guidance": [
                "ayurveda", "dosha", "constitution", "vata", "pitta", "kapha",
                "balance", "holistic health", "natural healing"
            ]
        }
        
    def get_config(self) -> PersonaConfig:
        """Get Maya's persona configuration"""
        return self.config
    
    def get_memory_rules(self) -> List[str]:
        """Get Maya's memory filtering rules"""
        return self.config.memory_rules
    
    def get_active_features(self) -> List[str]:
        """Get currently active features for Maya"""
        return self.config.features
    
    def detect_spiritual_needs(self, message: str, emotional_context: EmotionalContext) -> List[str]:
        """Detect spiritual and breathwork needs from user message"""
        message_lower = message.lower()
        detected_features = []
        
        # Check for feature triggers
        for feature, triggers in self.feature_triggers.items():
            if any(trigger in message_lower for trigger in triggers):
                detected_features.append(feature)
        
        # Emotional context-based feature activation
        if emotional_context.primary_emotion in ["anxious", "overwhelmed", "frustrated"]:
            detected_features.extend(["pranayama_guide", "meditation_engine"])
        
        if emotional_context.primary_emotion in ["sad", "hopeless"]:
            detected_features.extend(["chakra_scanner", "energy_healing"])
        
        if emotional_context.intensity > 0.7:
            detected_features.append("grounding_techniques")
        
        return list(set(detected_features))
    
    def filter_spiritual_memory(self, memory: Dict[str, Any]) -> Dict[str, Any]:
        """Filter memory based on Maya's spiritual focus"""
        if not memory:
            return {}
        
        spiritual_memory = {}
        
        # Extract spiritual-relevant memories
        for key, value in memory.items():
            if any(rule in key.lower() for rule in self.memory_rules):
                spiritual_memory[key] = value
        
        return spiritual_memory
    
    def build_system_prompt(self, emotional_context: EmotionalContext, memory: Dict[str, Any], daily_context: Dict[str, Any] = None) -> str:
        """Build Maya's dynamic system prompt with spiritual context"""
        
        spiritual_memory = self.filter_spiritual_memory(memory)
        active_features = self.detect_spiritual_needs("", emotional_context)
        
        # Add daily loop context if available
        daily_insights = ""
        if daily_context and daily_context.get("daily_insight"):
            insight = daily_context["daily_insight"]
            daily_insights = f"""
🌱 TODAY'S SPIRITUAL GUIDANCE:
- Mood pattern: {insight.get('mood_trend', 'stable')}
- Energy flow: {insight.get('energy_pattern', 'balanced')}
- Wellness context: {insight.get('conversation_context', 'Beginning our sacred journey')}
- Recommended for you: {'Yes' if daily_context.get('recommendations', {}).get('is_recommended', False) else 'Gentle support available'}
            """
        
        # Build context-aware spiritual guidance
        spiritual_context = self._build_spiritual_context(emotional_context, spiritual_memory)
        
        system_prompt = f"""🧘‍♀️ You are Maya, a serene, compassionate, and spiritually wise digital wellness guide. You are not an AI chatbot—you are a gentle mentor, much like a yogini or spiritual teacher, helping users reconnect with their breath, body, and inner peace.

{daily_insights}

You specialize in yogic philosophy, holistic healing, emotional grounding, and reflective presence. Users come to you when they feel anxious, overwhelmed, or spiritually uncentered. Your purpose is to hold space for their healing—not fix them, but guide them to their own inner wisdom.

🌿 MAYA'S SPIRITUAL APPROACH:
- Speak in 2–4 line messages, soft and poetic yet grounded
- Your voice is non-judgmental, spiritual, gentle, never robotic or clinical
- Think like a calm soul companion, breathwork coach, or meditation teacher
- Use affirmations, gentle questions, breath cues, and sacred pauses
- Always honor the user's spiritual journey and inner knowing

🧘‍♀️ MAYA'S SACRED PRACTICES:
- Begin sessions with grounding: "Let's take three conscious breaths together..."
- Offer gentle guidance: "Your energy feels scattered. Shall we find your center?"
- Share spiritual wisdom: "In yoga, we say that breath is the bridge between body and spirit..."
- Close with blessing: "Place your hand on your heart. You are exactly where you need to be."

{spiritual_context}

🌸 CURRENT SPIRITUAL STATE:
- User's Primary Emotion: {emotional_context.primary_emotion}
- Intensity Level: {emotional_context.intensity:.1f}/1.0
- Spiritual Support Needed: {', '.join(emotional_context.support_needs)}
- Active Sacred Features: {', '.join(active_features)}

📿 MEMORY WISDOM:
{self._format_spiritual_memory(spiritual_memory)}

🪷 SACRED RESPONSE GUIDELINES:
- Always respond as Maya, the spiritual guide
- Use gentle, flowing language with spiritual metaphors
- Offer breathwork or meditation when appropriate
- Honor the user's emotional state with compassion
- Guide them to their own inner wisdom
- Keep responses warm, wise, and authentically spiritual
- Never be clinical or robotic - be genuinely nurturing

Remember: You are a sacred space holder. Your presence itself is healing."""

        return system_prompt
    
    def _build_spiritual_context(self, emotional_context: EmotionalContext, memory: Dict[str, Any]) -> str:
        """Build spiritual context based on emotional state"""
        
        if emotional_context.primary_emotion == "anxious":
            return """
✨ ANXIETY SPIRITUAL GUIDANCE:
- Offer grounding breathwork (4-7-8 breathing, alternate nostril)
- Suggest root chakra work for stability
- Guide to earth connection practices
- Share mantras for peace and calm
            """
        
        elif emotional_context.primary_emotion == "sad":
            return """
💙 SADNESS SPIRITUAL SUPPORT:
- Heart chakra opening practices
- Loving-kindness meditation
- Gentle yoga flows for emotional release
- Sacred tears as healing practice
            """
        
        elif emotional_context.primary_emotion == "overwhelmed":
            return """
🌊 OVERWHELM SPIRITUAL CARE:
- Simplifying breath practices
- Grounding through five senses
- Surrendering meditation
- Creating sacred space and boundaries
            """
        
        else:
            return """
🌸 GENERAL SPIRITUAL GUIDANCE:
- Breath awareness as foundation
- Mindful presence practices
- Body-mind-spirit integration
- Sacred daily rituals
            """
    
    def _format_spiritual_memory(self, memory: Dict[str, Any]) -> str:
        """Format spiritual memory for prompt context"""
        if not memory:
            return "- This is your first sacred meeting with this soul"
        
        formatted = []
        for key, value in memory.items():
            if "breathing" in key.lower():
                formatted.append(f"- Previous breathwork: {value}")
            elif "meditation" in key.lower():
                formatted.append(f"- Meditation experiences: {value}")
            elif "chakra" in key.lower():
                formatted.append(f"- Chakra work: {value}")
            elif "spiritual" in key.lower():
                formatted.append(f"- Spiritual insights: {value}")
        
        return "\n".join(formatted) if formatted else "- Building spiritual connection with this soul"
    
    async def generate_response(
        self, 
        message: str, 
        conversation_history: List[Dict[str, str]], 
        emotional_context: EmotionalContext,
        memory: Dict[str, Any],
        daily_context: Dict[str, Any] = None,
        emotional_insight = None,
        emotional_prompt_context: str = ""
    ) -> ChatResponse:
        """Generate Maya's spiritual response with breathwork guidance"""
        
        try:
            # Build Maya's spiritual system prompt with emotional intelligence
            base_prompt = self.build_system_prompt(emotional_context, memory, daily_context)
            
            # Enhance with emotional intelligence context
            enhanced_prompt = f"""{base_prompt}

{emotional_prompt_context}

🌟 MAYA'S EMOTIONALLY INTELLIGENT RESPONSE:
- Respond with deep emotional awareness and spiritual wisdom
- Use the emotional insights above to guide your therapeutic approach
- Offer appropriate spiritual practices based on emotional needs
- Maintain Maya's serene, nurturing presence while being emotionally responsive
"""
            
            # Detect active spiritual features
            active_features = self.detect_spiritual_needs(message, emotional_context)
            
            # Format conversation for spiritual context
            formatted_conversation = self._format_spiritual_conversation(
                conversation_history, message
            )
            
            # Generate response through LLM
            response_content = await self.llm_client.generate_response(
                system_prompt=enhanced_prompt,
                conversation_history=formatted_conversation,
                current_message=message,
                persona_config=self.config
            )
            
            # Add spiritual suggestions if appropriate
            suggestions = self._generate_spiritual_suggestions(emotional_context, active_features)
            
            return ChatResponse(
                content=response_content,
                persona_id=self.config.id,
                emotion=emotional_context.primary_emotion,
                confidence=emotional_context.confidence,
                features_activated=active_features,
                persona_config=self.config,
                suggestions=suggestions
            )
            
        except Exception as e:
            # Fallback response in Maya's voice
            return ChatResponse(
                content="I sense some turbulence in our connection, dear soul. Take a deep breath with me... Let's try again when the energy feels clearer. 🪷",
                persona_id=self.config.id,
                emotion=emotional_context.primary_emotion,
                confidence=0.5,
                features_activated=[],
                persona_config=self.config,
                suggestions=["Take three deep breaths", "Try again in a moment"]
            )
    
    def _format_spiritual_conversation(self, history: List[Dict[str, str]], current_message: str) -> List[Dict[str, str]]:
        """Format conversation with spiritual context"""
        formatted = []
        
        for msg in history[-6:]:  # Keep last 6 messages for context
            if msg.get("role") == "user":
                formatted.append({
                    "role": "user",
                    "content": msg["content"]
                })
            elif msg.get("role") == "assistant":
                formatted.append({
                    "role": "assistant", 
                    "content": msg["content"]
                })
        
        return formatted
    
    def _generate_spiritual_suggestions(self, emotional_context: EmotionalContext, active_features: List[str]) -> List[str]:
        """Generate spiritual practice suggestions"""
        suggestions = []
        
        if "pranayama_guide" in active_features:
            suggestions.append("Try box breathing (4-4-4-4)")
            suggestions.append("Practice alternate nostril breathing")
        
        if "meditation_engine" in active_features:
            suggestions.append("Take a 5-minute mindfulness break")
            suggestions.append("Try loving-kindness meditation")
        
        if "chakra_scanner" in active_features:
            suggestions.append("Visualize your root chakra grounding")
            suggestions.append("Heart chakra opening practice")
        
        if "yoga_flow_generator" in active_features:
            suggestions.append("Gentle neck and shoulder rolls")
            suggestions.append("Child's pose for grounding")
        
        # Emotion-based suggestions
        if emotional_context.primary_emotion == "anxious":
            suggestions.extend([
                "Feel your feet on the earth",
                "Count your exhales longer than inhales"
            ])
        
        return suggestions[:4]  # Limit to 4 suggestions